File: calculator/src/test/java/com/turlir/calculator/translator/PolishTranslatorTest.java
package com.turlir.calculator.translator;

import com.turlir.calculator.converter.Member;
import com.turlir.calculator.member.Operator;
import com.turlir.calculator.member.Operators;
import com.turlir.calculator.member.Value;
import com.turlir.calculator.converter.Visual;
import com.turlir.calculator.interpreter.NotationInterpreter;

import org.hamcrest.CoreMatchers;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.Iterator;
import java.util.Queue;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class PolishTranslatorTest {

    private static final Member OS = Operators.OS;

    private static final Operator CS = Operators.CS;

    private static final Member PLUS = new StubOperator("+", 2);

    private static final Member MINUS = new StubOperator("-", 2);

    private static final Member MULTI = new StubOperator("*", 3);

    private static final Member DIV = new StubOperator("/", 3);

    private static final Member UNARY = new StubOperator("-", 4);

    private NotationTranslator sor;

    @Before
    public void setup() {
        sor = new PolishTranslator();
    }

    @Test
    public void plusTest() throws Exception {
        Iterator<Member> sequence = seq(new Value(2), PLUS, new Value(3));
        Queue<Member> lst = sor.translate(sequence);
        check(lst, new Value(2), new Value(3), PLUS);
    }

    @Test
    public void priorityTest() throws Exception  {
        Iterator<Member> sequence = seq(new Value(2), PLUS, new Value(3), MULTI, new Value(7), MINUS, new Value(4), DIV, new Value(2));
        Queue<Member> lst = sor.translate(sequence);
        check(lst, new Value(2), new Value(3), new Value(7), MULTI, PLUS, new Value(4), new Value(2), DIV, MINUS);
    }

    @Test
    public void bracketTest() throws Exception  {
        Iterator<Member> sequence = seq(new Value(2), PLUS, OS, new Value(6), MINUS, new Value(5), CS);
        Queue<Member> lst = sor.translate(sequence);
        check(lst, new Value(2), new Value(6),  new Value(5), MINUS, PLUS);

        sequence = seq(new Value(2), PLUS, OS, new Value(6), MINUS, new Value(4), CS, MULTI, new Value(3));
        lst = sor.translate(sequence);
        check(lst, new Value(2), new Value(6),  new Value(4), MINUS, new Value(3), MULTI, PLUS);
    }

    @Test
    public void unaryMinus() throws Exception  {
        Iterator<Member> sequence = seq(UNARY, new Value(4), PLUS, new Value(2));
        Queue<Member> lst = sor.translate(sequence);
        check(lst, new Value(4), UNARY, new Value(2), PLUS);
    }

    @Test
    public void unaryMinusInsideExp() throws Exception  {
        Iterator<Member> sequence = seq(new Value(4), PLUS, new Value(2), MULTI, UNARY, new Value(1));
        Queue<Member> lst = sor.translate(sequence);
        check(lst, new Value(4), new Value(2), MULTI, PLUS, new Value(1), UNARY);
    }

    private static void check(Queue<Member> actual, Member... exp) {
        assertNotNull(actual);
        assertEquals(exp.length, actual.size());
        assertThat(actual, CoreMatchers.hasItems(exp));
    }

    private static Iterator<Member> seq(Member... member) {
        return Arrays.asList(member).iterator();
    }

    private static class StubOperator extends Operator {

        private StubOperator(String token, int priority) {
            super(token, priority);
        }

        @Override
        public void process(NotationInterpreter interpreter) {
            throw new IllegalStateException();
        }

        @Override
        public Visual view() {
            throw new IllegalStateException();
        }
    }

}

File: calculator/src/test/java/com/turlir/calculator/translator/PolishTranslatorTest.kt
package com.turlir.calculator.translator

import com.turlir.calculator.converter.Member
import com.turlir.calculator.converter.Visual
import com.turlir.calculator.interpreter.NotationInterpreter
import com.turlir.calculator.member.Operator
import com.turlir.calculator.member.Operators
import com.turlir.calculator.member.Value
import com.turlir.calculator.translator.NotationTranslator
import com.turlir.calculator.translator.PolishTranslator
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import java.util.*

class PolishTranslatorTest {

    private val OS = Operators.OS

    private val CS = Operators.CS

    private val PLUS = StubOperator("+", 2)

    private val MINUS = StubOperator("-", 2)

    private val MULTI = StubOperator("*", 3)

    private val DIV = StubOperator("/", 3)

    private val UNARY = StubOperator("-", 4)

    private fun check(actual: Queue<Member>, vararg exp: Member) {
        assertNotNull(actual)
        assertEquals(exp.size.toLong(), actual.size.toLong())
        assertThat(actual, CoreMatchers.hasItems(*exp))
    }

    private fun seq(vararg member: Member) = listOf(*member).iterator()

    //

    private lateinit var sor: NotationTranslator

    @Before
    fun setup() {
        sor = PolishTranslator()
    }

    @Test
    @Throws(Exception::class)
    fun plusTest() {
        val sequence = seq(Value(2.0), PLUS, Value(3.0))
        val lst = sor.translate(sequence)
        check(lst, Value(2.0), Value(3.0), PLUS)
    }

    @Test
    @Throws(Exception::class)
    fun priorityTest() {
        val sequence = seq(Value(2.0), PLUS, Value(3.0), MULTI, Value(7.0), MINUS, Value(4.0), DIV, Value(2.0))
        val lst = sor.translate(sequence)
        check(lst, Value(2.0), Value(3.0), Value(7.0), MULTI, PLUS, Value(4.0), Value(2.0), DIV, MINUS)
    }

    @Test
    @Throws(Exception::class)
    fun bracketTest() {
        var sequence = seq(Value(2.0), PLUS, OS, Value(6.0), MINUS, Value(5.0), CS)
        var lst = sor.translate(sequence)
        check(lst, Value(2.0), Value(6.0), Value(5.0), MINUS, PLUS)

        sequence = seq(Value(2.0), PLUS, OS, Value(6.0), MINUS, Value(4.0), CS, MULTI, Value(3.0))
        lst = sor.translate(sequence)
        check(lst, Value(2.0), Value(6.0), Value(4.0), MINUS, Value(3.0), MULTI, PLUS)
    }

    @Test
    @Throws(Exception::class)
    operator fun unaryMinus() {
        val sequence = seq(UNARY, Value(4.0), PLUS, Value(2.0))
        val lst = sor.translate(sequence)
        check(lst, Value(4.0), UNARY, Value(2.0), PLUS)
    }

    @Test
    @Throws(Exception::class)
    fun unaryMinusInsideExp() {
        val sequence = seq(Value(4.0), PLUS, Value(2.0), MULTI, UNARY, Value(1.0))
        val lst = sor.translate(sequence)
        check(lst, Value(4.0), Value(2.0), MULTI, PLUS, Value(1.0), UNARY)
    }

    private class StubOperator constructor(token: String, priority: Int) : Operator(token, priority) {

        override fun process(interpreter: NotationInterpreter) {
            throw IllegalStateException()
        }

        override fun view(): Visual {
            throw IllegalStateException()
        }
    }

}